### PR TITLE
[Sampling] Add Mirostat v2 sampling

### PR DESCRIPTION
**Pull request by chuan932**
## Purpose

Add opt-in **Mirostat v2** sampling (`mirostat_mode` / `mirostat_tau` /
`mirostat_eta`) to the V1 sampler. Mirostat (Basu et al., 2020,
https://arxiv.org/abs/2007.14966) is a feedback controller that keeps the
per-token *surprise* (`-log2 p`, i.e. per-token cross-entropy in bits) near a
target `tau` by maintaining a per-request threshold `mu` updated every step:

```
mu <- mu - eta * (observed_surprise - tau)
```

This bounds the per-token surprise from above, which keeps decoding out of the
*confusion trap* — the incoherent, non-terminating output that plain
temperature/top-p sampling falls into at high temperature. Measured on
DeepSeek-V4-Flash at `temperature=1.8-2.0`, it cuts the share of requests that
run all the way to `max_tokens` by ~85% and halves generated tokens (numbers in
[Model evaluation](#model-evaluation)). At `temperature <= 1.5` it is inert by
design and leaves well-behaved decoding untouched.

**Default `mirostat_mode=0` is a no-op**, so existing behavior is unchanged.

Usage:

```python
SamplingParams(mirostat_mode=2, mirostat_tau=5.0, mirostat_eta=0.1,
               temperature=1.0, top_k=0, top_p=1.0)
```

Also exposed on the OpenAI `/v1/completions` and `/v1/chat/completions` request
bodies as `mirostat_mode` / `mirostat_tau` / `mirostat_eta`.

Closes the long-standing request in #5209 (Mirostat), which was auto-closed for
inactivity, not rejected.

## Not a duplicate

Searched open PRs/issues (`gh pr list --search "mirostat"`, `gh search issues
mirostat`). No open PR implements Mirostat. #5209 is a stale-closed feature
request with no implementation; this PR implements Mirostat v2 for the V1
sampler.

## Design notes (for reviewers)

- **State lives in the `Sampler`, not a `LogitsProcessor`.** The `mu` update
  needs the probability of the *sampled* token, which only exists after
  sampling; a pre-sampling logits processor cannot see it. This mirrors the
  existing `ThinkingBudgetStateHolder` pattern. `MirostatStateHolder`
  (`vllm/v1/sample/mirostat_state.py`) is created per `InputBatch`, synced via
  `BatchUpdate` (add/remove/move), attached to `SamplingMetadata`, and consumed
  in `Sampler.sample()`.
- **Truncation is argmax-invariant.** Mirostat v2 sets the logit of any token
  with surprise `> mu` to `-inf` and always keeps the argmax, so it composes
  with the rest of the random-sampling pipeline. It runs after temperature and
  the argmax-invariant logits processors (e.g. min_p), before top-k/top-p.
- **Decode throughput is unaffected (GPU-resident state).** `mu`/`tau`/`eta`
  are persistent device tensors; the per-step math (truncate + `mu` update) is
  fully on-device with **no host<->device sync**. Inactive rows carry
  `mu = +inf` / `eta = 0`, making the per-step math a masked no-op for them, so
  a single vectorized path covers the whole batch with static shapes (CUDA
  graph friendly). When no request uses Mirostat the holder short-circuits, so
  there is zero overhead on the common path.
- **Validation.** `mirostat_mode` must be `0` or `2`. When enabled it requires
  `temperature > 0`, `top_k in {0, -1}`, `top_p == 1.0` (Mirostat replaces
  top-k/top-p truncation), and finite `tau`/`eta > 0`. Because temperature is
  applied before Mirostat, the `mirostat_mode` docstring recommends
  `temperature ~= 1.0` (a much lower temperature drives surprise toward 0 and
  makes Mirostat approach greedy).

## Tests

- `pytest tests/v1/sample/test_mirostat.py -v` → **14/14 passed** (parameter
  validation; holder add/remove/swap/move bookkeeping; truncation keeps the
  argmax; `mu` updates in both directions).
- `ruff check` and `ruff format --check` clean on all changed files (ruff
  0.14.0, the pinned CI version).

## Model evaluation

Measured on DeepSeek-V4-Flash (8xH20, TP=8).

**Repetition / degeneration** (HumanEval/132 `is_nested`, `think`-high,
N=20/config, `max_tokens=8192`; a "loop" = `finish_reason==length` with high
tail-repetition):

| config | loop rate |
| --- | --- |
| greedy (`temperature=0`) | 35% |
| plain sampling (`temperature=1.0`, `top_p=1.0`) | 0% |
| Mirostat v2 (`mode=2`, `tau=5`, `temperature=1.0`) | 0% |

Any `temperature > 0` already breaks the greedy repetition attractor, so
Mirostat is not what fixes greedy loops here. Being explicit about the scope:
with a fixed `seed`, Mirostat (`tau=5`) at `temperature <= 1.5` produces
**byte-identical** output to plain sampling. `tau=5` bits targets perplexity
`2^5 = 32`; below that temperature the observed surprise stays under `tau`, so
`mu` grows monotonically and the truncation is inert. That is the intended
behavior — Mirostat should not perturb well-behaved decoding. Its measurable
benefit shows up once the distribution is hot enough to enter the *confusion
trap*.

**High-temperature stability (confusion trap)** — this is the regime where
Mirostat is not substitutable by temperature tuning. Full HumanEval (164
problems, `think`-high, `top_p=1.0`, `max_tokens=8192`, 30 concurrent):

| temperature | config | pass@1 | truncated at `max_tokens` | mean out tokens | total out tokens |
| --- | --- | --- | --- | --- | --- |
| 1.8 | baseline | 4.9% (8/164) | 135/164 (82%) | 7217 | 1.18M |
| 1.8 | Mirostat v2 (`tau=5`) | **12.2%** (20/164) | **20/164 (12%)** | **3018** | **0.49M** |
| 2.0 | baseline | 0.6% (1/164) | 146/164 (89%) | 7718 | 1.27M |
| 2.0 | Mirostat v2 (`tau=5`) | **1.2%** (2/164) | **24/164 (15%)** | **3335** | **0.55M** |

At `temperature >= 1.8` plain sampling degenerates into incoherence and burns the
entire budget — the *median* request hits `max_tokens`. Mirostat cuts the
truncation rate by ~85% (82% -> 12% and 89% -> 15%), reduces total generated
tokens by ~57%, and raises pass@1 by 2.5x / 2x respectively. End-to-end
wall-clock for the two sweeps dropped 43% and 46%.

Scope of the win, stated plainly: this is damage mitigation, not recovery.
`temperature=1.8` + Mirostat (12.2%) is still far below `temperature=1.0` plain
sampling (99.4% pass@1 on the same harness). The value is that workloads which
*must* run hot (diversity / creative sampling) get a bounded-surprise safety net
instead of unbounded degeneration, at roughly half the tokens.

**Decode throughput — lossless** (same engine, 30 concurrent, `max_tokens=3000`,
non-think, measured from `vllm:generation_tokens_total`):

| config | decode tok/s |
| --- | --- |
| baseline (no Mirostat) | 1902.3 |
| Mirostat v2 (enabled) | **1895.1** |

Δ = **-0.4%, within noise**. An earlier host-syncing prototype regressed decode
by ~16% at 30 concurrency; moving the `mu` state fully GPU-resident (no per-step
`.tolist()`/D2H) removed the regression.

## Limitations / follow-ups

- Only Mirostat **v2** is implemented (the common variant). v1 (which estimates
  the Zipf exponent per step) can be added later.
- Not validated with speculative decoding; the `mu` update assumes one sampled
  token per request per step. Recommend not combining with spec decode for now.
- Mirostat v2 cannot break a **high-confidence** repetition loop, and should not
  be advertised as a general anti-repetition fix. The repeated token is the
  argmax, which the truncation always keeps, and its surprise is near 0, which
  drives `mu` *upward* and loosens truncation further. The controller bounds
  surprise from above (confusion trap), not from below. For greedy-style loops,
  raise `temperature` or use a repetition penalty.

## Note

AI assistance was used to prepare this change. The submitter has reviewed every
line and is responsible for the change and its tests/evaluations.

### CODE REVIEW

**Comment by chuan932** [comment]
For context, related issues this connects to (not superseded by this PR):

**Demand for anti-repetition / adaptive sampling**
- #5209 — Mirostat / dynamic-temperature / quadratic sampling request (auto-closed for inactivity, not rejected); this PR implements the Mirostat v2 part.
- #7842 — `no_repeat_ngram` in `SamplingParams`.
- #15379 — Top-nσ sampling.

**Complementary approach (different layer, same symptom)**
- #52673 — `ThinkingLoopBreaker`: *detects* exact periodic repetition inside `<think>` and forces `reasoning_end` early. That is a detection/backstop mechanism; Mirostat is a *sampling-time* controller that bounds per-token surprise so the distribution is less likely to collapse into a loop in the first place. The two are complementary and can be used together — Mirostat reduces how often loops start, `ThinkingLoopBreaker`/`thinking_token_budget` cap the damage when one still does.

**Real-world loop/degeneration reports Mirostat can mitigate**
- #31856 (MiniMax-M2.1-NVFP4 repetition loops), #49063 (greedy repetition loop with hybrid-MoE LoRA), #40080 / #15752 (Gemma repetition loops).

Note these are `Related to`, not `Closes` — Mirostat is a mitigation/opt-in sampler, not a root-cause fix for any specific model bug.

**Comment by chuan932** [comment]
### How this relates to the anti-repetition mechanisms already in vLLM

To pre-empt the natural "how is this different from what we already have?" question:

vLLM's existing loop/repetition mechanisms are all **detect-then-stop**, acting *after* a loop has already formed:

- **`RepetitionDetectionParams`** (`SamplingParams.repetition_detection`, enforced in `vllm/v1/core/sched/utils.py::check_sequence_repetition`) detects an exact N-gram cycle and *finishes the request* (`FINISHED_REPETITION` / `stop_reason="repetition_detected"`).
- **#52677 (ThinkingLoopBreaker)** reuses that same detector inside the reasoning section and *forces the reasoning-end sequence* so the model still answers.
- `repetition_penalty` / presence / frequency penalties push down already-seen tokens, but bluntly (they penalize legitimate repetition too).

Mirostat v2 operates at a different point: it is a **sampling-time feedback controller** that bounds per-token surprise around a target (`tau`), which keeps the output distribution from collapsing into the low-entropy state that *produces* loops in the first place. So it is **complementary, not overlapping**:

- Mirostat reduces how often a loop starts.
- `repetition_detection` / `ThinkingLoopBreaker` remain the backstop for when one still does.

They compose cleanly and can be enabled together. This PR does not touch or duplicate any of the above; `mirostat_mode` defaults to `0` (no-op), so existing behavior is unchanged.

**Comment by chuan932** [comment]
Gentle ping @yewentao256 — this touches `vllm/v1/worker/gpu_input_batch.py` and the v1 sampler, which fall under your code-owner areas. Would you have a chance to review when you get time? The change is gated behind `mirostat_mode=0` (no-op by default), and decode throughput is unchanged (1895 vs 1902 tok/s at 30 concurrency). Happy to address any feedback. Thanks!

**Comment by chuan932** [comment]
Gentle ping @njhill — this touches the V1 sampler (`vllm/v1/sample/`) and `vllm/v1/worker/gpu_input_batch.py`, areas you own. Would you have a chance to take a look when you get time?

Quick summary to make review cheap:
- `mirostat_mode` defaults to `0`, so the whole feature is a no-op unless a request opts in; existing behavior is unchanged.
- State lives in a `MirostatStateHolder` consumed inside `Sampler` (mirroring `ThinkingBudgetStateHolder`) rather than a `LogitsProcessor`, because the `mu` update needs the *sampled* token, which only exists after sampling.
- Fully on-device: `mu`/`tau`/`eta` persist as GPU tensors and inactive rows are masked (`mu=+inf`, `eta=0`), so there is no per-step host↔device sync and shapes stay static for CUDA graph capture. Measured decode throughput at 30-way concurrency: **1895 tok/s with mirostat vs 1902 tok/s without** (0.4%, within noise).

Happy to address any feedback. Thanks!

**Comment by njhill** [comment]
Hi @chuan932, the v1 model runner is now deprecated, if you want to propose something it must be based on the v2 model runner.

Note that custom logit processors is one of the remaining gaps that we're working on for the v2 model runner. It might be that this best fits as one of those once they are ready.

**Comment by chuan932** [comment]
Thanks @njhill. Understood — I won't continue the V1-specific path.

Mirostat v2 is stateful: it truncates logits before sampling, then updates `mu` from the sampled token afterward. Therefore, it needs a post-sampling state update in addition to the custom logits processor itself.

Would you prefer that I wait for the V2 custom-logits-processor interface and rework this PR once it lands, or would a V2-native sampler/state-holder implementation be acceptable now? I'm happy to convert this PR to draft while waiting.

**Review by yewentao256 — COMMENTED** [review]
~Thanks for the work! We don't want to introduce new update to V1 now as it is deprecated~

Same idea to Nick, feel free to have a new PR for v2 later, close for now
